Update topic/RMQ (#6)
OI-wiki/i18n: docs/topic/rmq.md

## Introduction

RMQ is the abbreviation of [Range Maximum/Minimum Query](https://en.wikipedia.org/wiki/Range_minimum_query), which means the maximum (minimum) value of an interval.

In the following descriptions, the default initial array size is $n$ , and the default time complexity is $O($ data preprocessing $)-O($ single query $)$ .

## Monotonic stack

Since **OI wiki** already has a detailed [description](../ds/monotonous-stack.md) of monotonic stack, we will not get into details here.

Time Complexity: $O(m\log m)-O(\log n)$ 

Space Complexity: $O(n)$ 

## Sparse Table

Since **OI wiki** already has a detailed [description](../ds/sparse-table.md) of sparse table, we will not get into details here.

Time Complexity: $O(n\log n)-O(1)$ 

Space Complexity: $O(n\log n)$ 

## Segment Tree

Since **OI wiki** already has a detailed [description](../ds/seg.md) of segment tree, we will not get into details here.

Time Complexity: $O(n)-O(\log n)$ 

Space Complexity: $O(n\log n)$ 

## Four Russian

The [Method of Four Russians](https://en.wikipedia.org/wiki/Method_of_Four_Russians) is an algorithm based on ST table proposed by four Russian computer scientists.

> The improvement made by the Four Russians algorithm based on the ST table is sequence blocking.

Reviewer: ```suggestion
The improvement made by the Four Russians algorithm, which is based on the ST table, is dividing a sequence into multiple blocks.
```
适当的使用插入语(其实某种意义上来说是他的逗号)来让句子修饰的内容更清晰
序列分块想了想还是拆成短句吧

Specifically speaking, we divide the original array, let's call it array A, into blocks of every $S$, for a total of $n/S$ blocks.

> For each block, we precomputate it to get the minimum value of the block, then create an array B with a length of $n/S$, and precomputate the array B using the ST table.

Author: 改成了 `..preprocess the elements within to get the minimum value..`？ :thinking: :

At the same time, we also create a ST table for each fragmented block of array A.

When accessing the array, we can divide the query interval into no more than one continuous block interval of array B and no more than two continuous intervals within the entire block of array A. Obviously, these problems can be solved by interval query in the ST table.

When $S=\log n$ , the precomputation complexity reaches the optimal, $O((n / \log n)\log n+(n / \log n)\times\log n\times\log \log n)=O(n\log \log n)$ 。

Time complexity $O(n\log \log n)-O(1)$ 

Space complexity $O(n\log \log n)$ 

Of course, because we need to run three ST tables, the constant of this method is relatively large.

!!! note "Some minor algorithm optimizations"
    We found that when the two endpoints of the query belong to different blocks in array A, the query in the blocks of array A is about the prefix or suffix of each block.

    Obviously, these queries can be solved within the time complexity of $O(n)$ by preprocessing the answers.

    In this way, we only need to perform the query operation on the ST table at most once.

!!! note "Some algorithm optimizations[ONLY FOR REFERENCE]"

>     Because the Four Russians algorithm is based on the ST table, and the algorithm contests does not generally have a very strict requirement about time complexity, it can generally be replaced by the ST table, which is not practical in the algorithm competition. Here we offer an improved version that is more practical in the contests.

Reviewer: > Because the Four Russians algorithm is based on the ST table, and the algorithm contests does not generally have a very strict requirement about time complexity, it can generally be replaced by the ST table, **which is not practical in the algorithm competition**. Here we offer an improved version that is more practical in the contests.
这里表述有点问题，这样说的话 `not practical` 修饰的是 `ST table`

>     We set the block size to $\sqrt n$, precompute the RMQ of the prefix and suffix within each block, and then brute force to precompute the RMQ between any consecutive blocks. For this solution, the time complexity is $O(n)$ .

Reviewer: ```suggestion
    We set the block size to $\sqrt n$, precompute the RMQ of the prefix and suffix within each block, and then use brute force to precompute the RMQ between any consecutive blocks. For this solution, the time complexity is $O(n)$ .
```

>     When querying, for those where the left and right endpoints are not in the same block, we can directly get the suffix RMQ of the block where left endpoint is located, the RMQ of the continuous block between the left and right endpoints, and the prefix RMQ of the block where right endpoint is located. The answer is the max/min value among the three.

Reviewer: **for those where** 语法有一些问题
建议 `for those queries which`，这样 `endpoints` 更清晰？

>     For queries of the block with both left and right endpoints, we can brute-force to find the RMQ between the two points. The time complexity is $O(\sqrt n)$, but the expected value of a single query to the block containing both left and right endpoints is $ O(\frac{\sqrt n}{n})$, so the time complexity of this method is the expected to be $O(n)$.

Reviewer: 这里所有的 `block` 显然都有左右端点，应和上一行一致？
另外建议 `expected time complexity` 因为这里的期望其实是数学期望（显然？）

>     In the algorithm competition, we don’t have to worry about the person who gets stuck getting this algorithm, because we can fine-tune the block size randomly on the basis of $\sqrt n$, which largely prevents the algorithm from being constructed according to a specific block size. The worst case appears in the data. And if the person who made the question wants to get rid of this method, the violence may pass.

Reviewer: > In the algorithm competition, we don’t have to worry about the **person who gets stuck getting this algorithm (1)**, because we can fine-tune the block size randomly **on the basis (2)**of $\sqrt n$, which largely prevents the algorithm from being constructed according to a specific block size. **The worst case appears in the data. (3)** And if the person who made the **question (4)** wants to **get rid of this method (5)**, the **violence (6)** may pass.

1. 这里翻译得有点迷惑
2. 好像不是问题太大 但是我标了号就不删了
3. 和上下文不太连贯
4. `problem` suggested
5. does not want to make solutions using this algorithm to be accepted，即“卡掉”可能要统一一个翻译（`stuck` 其实也并不是一个很好的翻译，歪果仁不一定了解这个东西）
6. brute force

    This is an algorithm that expects the time complexity to reach the lower bound, and the difficulty of the code implementation and time complexity constant are relatively small. So it is more practical in the algorithm competitions.

>     Note: The above algorithm refers to the solution in [P3793 Yuno rescues grandpa] (https://www.luogu.com.cn/problem/P3793)(original post in Chinese).

Reviewer: ```suggestion
    Note: The algorithm above refers to the solution of [P3793 Yuno rescues grandpa](https://www.luogu.com.cn/problem/P3793) (original post in Chinese).
```
`[]` 和 `()` 分开会导致一些 Markdown Parser 不认为这是个链接

## The application of Cartesian tree on RMQ

If you don't really know about Cartesian tree, please check out the [description here](../ds/cartesian-tree.md) 。

> We find that the min/max value between two points on the original sequence is equal to the LCA weight of the two points on the Cartesian tree, which means that what we need to solve now is find the LCA between two points on the $O(n)-O(1)$ tree.

Reviewer: 两个点 指的是节点 node

> How to find the LCA of the tree already has a description in the [LCA](../graph/lca.md) section, so we will not get into details here.

Reviewer: LCA 是关于树上两节点之间的运算，因此是 on a tree

What we need to use here is the LCA algorithm based on RMQ.

> Some of you might be wondering: wait, why do we circle back to the RMQ question again?

Reviewer: ```suggestion
Some of you might be wondering: wait, why do we turn back to the RMQ question again?
```

Don't worry, let's find out the special nature of this RMQ problem:

Because the two adjacent nodes of the tree's dfs traversal sequence have the parent-child relationship, the depth difference between the two adjacent nodes is $\pm 1$. We generally refer to this type of the RMQ problem where the difference between two adjacent elements is 1 as the $\pm 1$ RMQ problem.

> Based on this characteristic, we can now improve the Four Russians algorithm.

Reviewer: 特性，性质的用词统一一下 feature 吧

Since the bottleneck of the Four Russian algorithm is the RMQ problem within the blocks, we focus on the optimization of it.

Because the difference between two adjacent numbers is $\pm 1$, the number of the right sequence types whose length does not exceed $\log n$ when fixing the left endpoint number is $\sum_{i=1}^{i \ leq \log n} 2^{i-1}$, and this formula obviously does not exceed $n$.

> This suggests that we can precompute the minimum value of all cases that do not exceed $n$, that is, the value of the first element.

Reviewer: 这里是减号（
所以两侧有个空格

During precomputation, we need to precompute the difference between two adjacent numbers in the same block, and use binary to represent it.

When querying, we can find the binary representation corresponding to the query interval, and look up the answer in the table.

In this way, the time complexity of Four Russians precomputation is optimized to $O(n)$.

> Combining the Cartesian tree part, we can realize the RMQ problem of $O(n)-O(1)$.

Reviewer: ```suggestion
Combining the Cartesian tree part, we can realize the RMQ problem in $O(n)-O(1)$.
```

时间复杂度的介词统一用 in

The code and sample questions have already been given [here] (../graph/lca.md) in the LCA section, so we won't repeat it here.

Of course, due to the large number of conversion steps, the $O(n)-O(1)$ RMQ runs relatively slower.

> If the data is random, we can also do a brute-force search on the Cartesian tree. The time complexity is the expected value of $O(n)-O(\log n)$, and the complexity constant of this algorithm in the actualy implementation is often small.

Reviewer: 关于期望的修改同上

Author: 期望值的化应该就是 [expected value](https://en.wikipedia.org/wiki/Expected_value)?

RMQ is the abbreviation of [Range Maximum/Minimum Query](https://en.wikipedia.org/wiki/Range_minimum_query), which means the maximum (minimum) value of an interval.

> In the following descriptions, the uniform standard of initial array size is $n$ , and the default time complexity is $O($ data preprocessing $)-O($ single query $)$ .

Reviewer: 这种约定性的东西还是用 We assume that ... 吧 :thinking:
